Decouple pbench-move-results from server

ndokos opened the pull request:
pbench-move-results knows how to query a server to get a destination
for the tarballs that it has to move. It then checks that it can get
to the destination. If that is successful it packages up a tarball,
and calculates an md5. These two are copied to the destination. It
checks that the tarball's md5 is correct and if so, it marks the
transaction complete by renaming the md5 file at the destination.

This implementation will apply to version-002 clients. Version-001
clients will continue to do things as they do today.

pbench-move-results gets a new --user option.  The "user = <value>"
option gets added to the metadata.log in the [run] section before the
tarball is created. Its default value comes from the env variable
PBENCH_USER (if it exists), but may be overridden on the command line.

Any prefix specified on the command line is also handled the same way:
the "prefix = <value>" option gets added to the [run] section of
the metadata log.

The actual handling of both of these options is TBD: a future
server-side PR will implement the creation of prefix/user links to the
tarball.

Tarballs and md5 sum files are created in a temp directory
which is copied (recursively) to the server. The md5 file is
called <resultname>.tar.xz.md5.check and is renamed to omit
the .check suffix after a successful md5 check, thus signalling
completion of the moving and checking of the tarball.

The number of ssh invocations has thus been reduced to one initial
one to check connectivity plus two for each tarball: one to copy
the temp directory over and one to check the MD5 sum and rename
the file.

Duplicate detection has been eliminated from the agent side.
It is done on the server side (see below).

pbench-server-prep-shim-002 is a new script: it will scour the
reception directory for xxx.tar.xz.md5 files. For each file, it will
check the md5 sum and, if successful, it will copy the corresponding
tarball and its md5 to the archive directory and then create the link
in the TODO state directory. That will get the ball rolling on the
rest of the processing.  Errors like missing tarball, missing md5,
duplicate names, are detected and handled appropriately (mostly by
quarantining the tarball).

This takes care of version-002 clients. In a future PR, the
server-side handling of version-002 clients will change:

A new pbench-server-prep script will be added to do similar things for
the inotify version of the server: move the files into the archive and
add an entry to the dispatch list file that inotify is monitoring.
That will get *that* ball rolling. In fact, there will be two: one for version 001
clients and one for version 002 clients.

TBD: pbench-server-activate-setup-results-host-info now puts receive-dir
(which is defined in the server config file as
/pbench-local/fs-version-001/pbench-move-results-receive) into the
version-002 pbench-results-host-info file on the server, instead of the
archive-dir.

tenstormavi commented on agent/util-scripts/pbench-move-results:
```diff
@@ -137,17 +148,20 @@ if [ $? -ne 0 ]; then
 let failures=0
 
-trap "rm -f $pbench_tmp/prefix.*" EXIT INT QUIT
+tmp=${pbench_tmp}/${script_name}.$$/
```
There should not be a forward slash at the end because it is creating the wrong URL for the result.

tenstormavi commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
+    # If it does not fail, then check the MD5 sum and rename the foo.tar.xz.md5.check file
+    # to foo.tar.xz.md5. That's the signal that the agent has finished with this tarball.
+
+    tarball="$tmp/$controller/$pbench_run_name.tar.xz"
```
This URL will be wrong because of above forward slash in tmp.

portante replied:
@tenstormavi, you mean the there will be a 2 slash characters in a row in the URL?

Where will `$tarball` get used as a URL?

tenstormavi commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
+    tarballmd5="$tmp/$controller/$tarball.md5"
+    # we need to calculate the md5 sum in the temp directory
+    # in order to get the filename right.
+    pusdh $tmp/$controller
```
Here is a typo, this should be `pushd`.

portante replied:
For consistently, could you use `pushd $(dirname $tarball)` instead?

tenstormavi commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
     fi
-    md5sum "$tarball" > "$tarball.md5"
+
+    tarballmd5="$tmp/$controller/$tarball.md5"
```
This should be `tarballmd5="$tmp/$controller/${tarball##*/}.md5"` otherwise it is creating wrong URL of the md5.

ndokos commented on server/pbench/lib/config/pbench-server.cfg.example:
```diff
@@ -19,5 +20,5 @@ roles = pbench-results, pbench-backup
 [config]
-paths = /opt/pbench-server/lib/config
-files = pbench-server-default.cfg
+paths = /opt/pbench-server/lib/config, /home/nick/src/github/pbench/worktrees/wip-pbench-move-results/server/pbench/lib/config
```
Thanks for the review.

ndokos commented:
@portante, @tenstormavi, @chaitanyaenr: please rreview this! I believe I've fixed the things that @tenstormavi pointed out, but it needs more eyes (plust a unit test for the shim).

portante commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
+    pusdh $tmp/$controller
+    md5sum "$(basename $tarball)" > "$tarballmd5.check"
+    popd
     if [ $? -ne 0 ]; then
```
Is this going to check the status of `popd` instead of `md5sum`?

portante replied:
This question still stands ...

ndokos replied:
Yes - new version coming.

portante commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
     if [ $? -ne 0 ]; then
         error_log "ERROR: md5sum failed for $tarball, skipping"
-        rm -f $tarball $tarball.md5
+        rm -f "$tarball" "$tarballmd5"
```
This seems to be missing `.check` for the MD5 file name?

portante commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
-        error_log "ERROR: unable to copy results tarball, $tarball, to $results_repo:$results_full_path"
-        rm -f $tarball $tarball.md5
+        error_log "ERROR: unable to copy results tarball, $tarball, to $results_repo:$results_path_prefix"
+        rm -f $tarball $tarballmd5
```
`.check`

ndokos replied:
I changed the definition of `tarballmd5` to included the `.check`. There is only one place where the `.check` should not be there.

portante commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
-
     # Verify the bits copied are good
-    ssh -i $pbench_bin/id_rsa $ssh_opts $results_repo "cd $results_full_path; md5sum --check $pbench_run_name.tar.xz.md5"
+    md5name=$pbench_run_name.tar.xz.md5
```
Could this be `$(basename $tarballmd5)`?

ndokos replied:
Changed it to `$(basename $tarball).md5` instead.

portante commented on agent/util-scripts/pbench-move-results:
```diff
@@ -172,120 +186,84 @@ for dir in `/bin/ls -ort -d */ | awk '{print $8}' | grep -v "^tools-" | grep -v
     if [ $chk_res -ne 0 ]; then
         error_log "ERROR: remote copy failed, remote tarball MD5 does not match original"
-        rm -f $tarball $tarball.md5
+        rm -f $tarball $tarballmd5
```
`.check`

portante replied:
This looks fine now ...

portante commented on server/pbench/lib/config/pbench-server.cfg.example:
```diff
@@ -4,6 +4,7 @@
 # An installation will want to override at least the following:
 mailto = admins@example.com
 default-host = pbench.example.com
+version=002
```
Do we want to expose the version number as a configuration variable?

ndokos replied:
No - deleted. The default config file will have a `versions = 001 002` def to indicate that the server handles both (it will also be needed in the installation to create the host-info-url files).

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+#! /bin/bash
+
```
Can we add a short description of that this script does here?

ndokos replied:
Yes.

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+    dest=$1
+    files="$@"
+
+    mkdir -o $dest || return $?
```
s/`-o`/`-p`/

ndokos replied:
Off by one error...

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+tmp=$TMP/$PROG.$$
+
+function quarantine () {
+    dest=$1
```
Do we need to `shift` after referencing `$1` so that `$@` doesn't include the `dest` value?

ndokos replied:
Yes.

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+    resultname=$(basename ${tb})
+    resultname=${resultname%.tar.xz}
+
+    # hostname is the last component of the directory part of the full path
```
Perhaps we could say, "`# controller (hostname) is the last ...`".

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+
+    dest=$archive/$controller
+
+    # if there is a prefix file with the old name, normalize it now
```
Given this is supporting new clients, why would we have a prefix file with an old name here?

ndokos replied:
We would not - I got rid of this earlier, but I guess I never pushed the change.

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+        fi
+    fi
+
+    # sanity checks
```
Can we check the tarball before we do the prefix handling above?

ndokos replied:
No more prefix handling in this script.

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+    # sanity checks
+    if [ ! -f ${tb} ] ;then
+        echo "$TS: ${tb} file does not exist" >&4
+        quarantine $quarantine/$controller ${tb}.md5 $prefixname
```
Should this be:

```
quarantine $quarantine/$controller ${tb}.md5 ${tb}.prefix
```

ndokos replied:
It is now just
 
     echo "$TS: Processed $ntbs tarballs,   
         $nquarantined quarantined tarballs, $ndups duplicately-named tarballs." | tee -a $status

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+
+    if [ -f ${dest}/${resultname}.tar.xz -o -f ${dest}/${resultname}.tar.xz.md5 ] ;then
+        quarantine $duplicates/$controller  ${tb} ${tb}.md5 $prefixname
+        ndups=$ndups+1
```
Should we then `continue` after incrementing the duplicate count?

portante commented on server/pbench/bin/pbench-server-prep-version-001:
```diff
@@ -0,0 +1,182 @@
+    if [ $sts -eq 0 ] ;then
+        #if the mv succeeded, then delete the check file.
+        # otherwise, try again next time.
+        rm -f ${tb}.check
```
Delete the check file?  Not sure what that means...

ndokos replied:
Sorry - the last 3 comments refer to outdated code. I'll push the current code.

ndokos commented:
@portante: PTAL.

Still missing a unit test for the shim and I need to reshuffle/squash commits (I left it as is in the hope that the review will be easier, although the version of the shim in the base commit is different from the one you reviewed above - sorry about that.)

ndokos commented:
@portante: pushed "final" changes, squashed into two commits: one for code, one for unit tests, PTAL.

portante commented:
@ndokos should I still see 10 commits?

ndokos commented:
Ah, yes, I was fixing things without squashing commits. Let me do some squashing.

portante commented on server/pbench/bin/pbench-server-activate-setup-results-host-info:
```diff
@@ -46,10 +46,10 @@ if [ ! -z "$testdir" ] ;then
+# the directory and filename are specified in the agent config file as
+# the value of [results]/host_info_url and the setting is duplicated
+# (except for the version number which is appended here)in the server
+# config file: if it changes in one, it *has* to change in the other.
```
Do you remember why we need to have this value in two places?

ndokos replied:
One is in the agent and one is in the server config file. The server activation creates the file but the agent has to know where it is.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+
+tmp=$TMP/$PROG.$$
+
+function quarantine () {
```
Can we use the quarantine definition from `pbench-base.sh`?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+status=$tmp/status
+> $status
+# we are explicitly handling version-001 data
+results_dir=$(getconf.py receive-dir-${version} pbench-server)
```
This appears to be the only place we reference `${version}`.  Can we consider getting rid of that variable and using `receive-dir-001` instead?  We already reference 001 in other comments, so the variable does not appear to be that useful.

ndokos replied:
OK

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+        continue
+    fi
+
+    # This should not happen: version 001 agents check for duplicates in pbench-move-results.
```
Perhaps reword this comment to say, "While version 001 agents check for duplicates in their copy of `pbench-move-results`, they are only able to check against their 'archive` location, not against the real archive location.  So this check is still necessary to avoid actual duplicates."

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+        nerrs=$nerrs+1
+    fi
+
+    mkdir -p $dest/TODO
```
Can you move this one to line 170 so that we can just do one `mkdir -p`?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: ${tb} failed MD5 check" >&4
+        quarantine $quarantine/$controller ${tb} ${tb}.md5
+        sts=$?
```
Can we move all this quarantine error handling into the `quarantine` function in pbench-base.sh, and then have that error handling exit the process entirely when errors are encountered?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+
+    mv ${tb} ${tb}.md5 ${prefix} $dest
+    sts=$?
+    if [ $sts -ne 0 ] ;then
```
Can we quarantine when this `mv` command fails?

ndokos replied:
Absolutely.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,220 @@
+    fi
+
+    cd $dest/TODO > /dev/null
+    ln -s $dest/$(basename $tb) .
```
When this `ln` command fails, can we quarantine from the archive directory?

ndokos commented:
@portante: done - PTAL.

ndokos commented:
@portante: Spoke too soon - another version coming. Please wait.

ndokos commented:
@portante: OK - all done. PTAL.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -131,47 +133,40 @@ while read ck ;do
+    mkdir -p ${dest}/TODO
     sts=$?
     if [ $sts -ne 0 ] ;then
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5
```
Can we emit the error message before the quarantine?

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -131,47 +133,40 @@ while read ck ;do
     mv ${tb} ${tb}.md5 $dest
     sts=$?
     if [ $sts -ne 0 ] ;then
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5
```
Can we emit the error message before the quarantine?

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -131,47 +133,40 @@ while read ck ;do
     ln -s $dest/$(basename $tb) .
     sts=$?
     if [ $sts -ne 0 ] ;then
+        quarantine ${errors}/${controller} ${dest}/${tb} ${dest}/${tb}.md5 ${prefix}
```
Can we emit the error message before the quarantine?

ndokos commented:
@portante: all set I think. All: please review away!

ndokos commented:
@portante, @tenstormavi, @ravi: all squashed down and ready for review.

tenstormavi commented:
@ndokos @portante Don't you think we should have error checks on where we are doing quarantine the tarball and its md5 for whatever reason. What if the quarantine itself fails? 

Other than this everything seems to be fine to me.

ndokos commented on server/pbench/bin/pbench-base.sh:
```diff
@@ -125,27 +124,12 @@ function log_finish {
 
-    flock -n $log_file cat $TMP_DIR/$(basename $0).log >> $log_file
-    flock -n $error_file cat $TMP_DIR/$(basename $0).error >> $error_file
+    mkdir -p $dest || exit 101
```
Should the errors go to a log somewhere?

ndokos commented:
@tenstormavi suggested that the quarantine function log its errors. I implemented that suggestion in the latest commit. I also fixed a long-standing problem in pbench-move-results that @portante had pointed out but I missed the comment before: basically, a status check was checking the status of a `popd` instead of the preceding `md5`. Please take a look. I squashed the latter change. I'll squash the former if it passes muster.

tenstormavi commented on server/pbench/bin/pbench-base.sh:
```diff
@@ -124,12 +124,30 @@ function log_finish {
+# context.  Errors here are fatal but we log an error message to help
+# diagnose problems.
 function quarantine () {
+    set -x
```
I guess this should be commented out to pass the unit tests.

ndokos replied:
Yes - thanks! I'll squash everything down to the standard two commits later on today.

ndokos commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+# agents create a link in the TODO subdir (which is created if
+# necessary) after they finish with MD5 checking, so that's what
+# we look for.
+find ${results_dir} -maxdepth 3 -type l -name '*.tar.xz' > $list
```
No, it really should be `-type l`: version 001 agents copy the stuff, check the MD5 and then create a link in the TODO subdir of the destination. BTW, the extra subdir level is why the maxdepth is 3.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    exit 1
+fi
+
+if [[ "$quarantine" == "" ]] ;then
```
Won't `$quarantine` never be an empty string?

ndokos replied:
Ah, right, I should be checking $qdir. I'll fix.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+errors=${qdir}/errors-001
+
+errlog=$LOGSDIR/$PROG/$PROG.error
+if [[ "$archive" == "" ]] ;then
```
Can we move this check to the line immediately following the config fetch?

ndokos replied:
I'll fix.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    exit 1
+fi
+
+if [[ "$duplicates" == "" ]] ;then
```
Won't `$duplicates` never be an empty string?

ndokos replied:
Yes, if we check $qdir, this check can go away completely.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+###########################################################################
+
+archive=$(getconf.py archive-dir pbench-server)
+qdir=$(getconf.py quarantine-dir pbench-server)
```
Can we add a check for `$qdir` immediately following this config fetch?

ndokos replied:
Yes.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+typeset -i nquarantined=0
+typeset -i ndups=0
+
+echo $TS
```
Given the `find` command can take a long time, should we move this `echo` of the start timestamp up before we issue the `find`?

ndokos replied:
TS is set by pbench-base.sh once and for all. The only thing that moving it will accomplish will be to add it to the log file earlier. I can do that but unless somebody is actively looking at the log file, it won't make any difference.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+status=$tmp/status
+> $status
+# we are explicitly handling version-001 data
+results_dir=$(getconf.py receive-dir-001 pbench-server)
```
Can we move this up before the `log_init` call with the other configuration parameters are checked, and add a check that we get a value?

ndokos replied:
OK

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+# agents create a link in the TODO subdir (which is created if
+# necessary) after they finish with MD5 checking, so that's what
+# we look for.
+find ${results_dir} -maxdepth 3 -type l -name '*.tar.xz' > $list
```
Should we be checking the success/failure of the `find` command itself, and if the contents of the list are empty or not?

ndokos replied:
We should...

ndokos replied:
Actually, an empty list is OK: the while loop is skipped.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    fi
+    popd >/dev/null 2>&1
+
+    # sanity checks
```
Shouldn't these checks be performed immediately after we fetch the symlink value above, before we handle the prefix?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+
+    dest=$archive/$controller
+
+    pushd ${tbdir} >/dev/null 2>&1
```
Can we move this prefix handling work down to where we also check the `md5sum`, so that we can `pushd` once after we have verified the directory exists?

ndokos replied:
I preferred to do the prefix handling early, so that if something goes wrong, the new-style prefix can be quarantined. But that does require two pushd/popd sections. Let me know which way you want to go.

portante replied:
Perhaps drop the `pushd/popd` and use `${tbdir}` in all the file name references?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    if [ ! -f ${tb} ] ;then
+        echo "$TS: ${tb} file does not exist" | tee -a $status >&4
+        if [ -f ${tb}.md5 ] ;then
+            quarantine ${quarantine}/${controller} ${tb}.md5 ${prefix}
```
Should we count the number of times we quarantined data?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+
+    if [ ! -f ${tb}.md5 ] ;then
+        echo "$TS: ${tb}.md5 file does not exist" >&4
+        quarantine ${quarantine}/${controller} ${tb} ${prefix}
```
Should we count the number of times we quarantined data?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    # script gets to them and moves them.
+    if [ -f ${dest}/${resultname}.tar.xz -o -f ${dest}/${resultname}.tar.xz.md5 ] ;then
+        echo "$TS: ${tb} duplicate name" >&4
+        quarantine $duplicates/$controller  ${tb} ${tb}.md5 $prefix
```
Should we count the number of times we quarantined data?

ndokos replied:
This is counted as a dup.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: Failed: \"mkdir -p ${dest}\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5 $prefix
```
Should we count the number of times we quarantined data?

ndokos replied:
This is counted as an error only. Basically, the (perhaps misnamed) function quarantine moves stuff to different destinations: quarantine, duplicates or errors.

portante replied:
As we talked over the wall, perhaps have the counts add up properly so that we can tell what actions were taken by the report only?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: Failed: \"mv ${tb} ${tb}.md5 ${prefix} $dest\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5 $prefix
```
Should we count the number of times we quarantined data?

ndokos replied:
This is counted as an error only.

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+        fi
+        echo "$TS: Failed: \"ln -s $dest/$(basename $tb) .\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${dest}/${tb} ${dest}/${tb}.md5 ${prefix}
```
Should we count the number of times we quarantined data?

ndokos replied:
This is counted as an error only.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+        continue
+    fi
+
+    pushd $(dirname $tb) >/dev/null 2>&4
```
s/$tb/$tbdir/

ndokos replied:
pushd ${tbdir}, right?

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+    fi
+
+    pushd $(dirname $tb) >/dev/null 2>&4
+    md5sum --check $(basename $tb).md5
```
s/$(basename $tb)/$resultname/

ndokos replied:
Right.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+        echo "$TS: ${tb} duplicate name" >&4
+        quarantine ${duplicates}/${controller}  ${tb} ${tb}.md5
+        ndups=$ndups+1
+        nerrs=$nerrs+1
```
Are duplicates errors?  If so, we should count them as errors here and in the `-shim-001` script as well.  If let's drop the counting here.

ndokos replied:
I think I'm going to count them as errors.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+        if [ -f ${tb}.md5 ] ;then
+            quarantine ${quarantine}/${controller} ${tb}.md5
+        fi
+        nquarantined=$nquarantined+1
```
Should we only count quarantines when we call the `quarantine` function?

ndokos replied:
Moved the counting inside the if (but as mentioned before, not every call to quarantine increments the quarantine count).

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+# Check for results that are ready for processing: version 002 agents
+# upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
+# after they are done with MD5 checking so that's what we look for.
+find ${results_dir} -maxdepth 2 -name *.tar.xz.md5 > $list
```
Same comments from the beginning to here as in `-shim-001` ...

ndokos replied:
Yes.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+
+    if [ -f ${dest}/${resultname}.tar.xz -o -f ${dest}/${resultname}.tar.xz.md5 ] ;then
+        echo "$TS: ${tb} duplicate name" >&4
+        quarantine ${duplicates}/${controller}  ${tb} ${tb}.md5
```
Should we count the number of times we quarantined data?

ndokos replied:
This counts as a duplicate (and an error).

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: Failed: \"mkdir -p ${dest}\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5
```
Should we count the number of times we quarantined data?

ndokos replied:
This counts as an error only.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: Failed: \"mv ${tb} ${tb}.md5 $dest\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${tb} ${tb}.md5
```
Should we count the number of times we quarantined data?

ndokos replied:
This counts as an error only.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+    if [ $sts -ne 0 ] ;then
+        echo "$TS: Failed: \"ln -s $dest/$(basename $tb) .\", status $sts" |
+            tee -a $status >&4
+        quarantine ${errors}/${controller} ${dest}/${tb} ${dest}/${tb}.md5 ${prefix}
```
Should we count the number of times we quarantined data?

ndokos replied:
This counts as an error only.

portante commented on server/pbench/bin/pbench-server-prep-shim-002:
```diff
@@ -0,0 +1,175 @@
+
+log_finish
+
+if [ $nerrs -gt 0 ] ;then
```
Can we issue the report regardless of errors?

portante commented on server/pbench/bin/pbench-server-prep-shim-001:
```diff
@@ -0,0 +1,213 @@
+
+log_finish
+
+if [ $nerrs -gt 0 ] ;then
```
Can we issue the report regardless of the number of errors?

portante commented:
And I seems like this needs a rebase as well.  Thanks.

ndokos commented on server/pbench/bin/pbench-server-activate-create-crontab:
```diff
@@ -85,7 +104,7 @@ for role in $roles ;do
     fi
     case $role in
-        pbench-rsync-satellites)
+        pbench-sync-satellites)
```
Not a bug fix, but it is munging different things into this commit. I should probably pull it out into a different PR. Basically, I reworked the init scripts in conjunction with the config files to allow easier specification of satellite syncing.

ndokos commented:
@portante: I think I've addressed all the review comments - thanks for the detailed review. I've rebased as well. PTAL.

ndokos commented:
@portante: hold off on the review - I've got to fix the unit tests.

portante commented on server/pbench/bin/pbench-dispatch:
```diff
@@ -148,7 +148,7 @@ else
 
         # move any prefix file to the .prefix subdir
```
Why is this logic here?

ndokos commented:
This is ready for (final?) review. I added the latest changes as extra commits but I'm going to squash them down to two for final merge. If it would make it easier to review, I could squash them now.

portante commented:
@ndokos, can you rebase and squash them?
